File: 2020/5.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "prova.txt"
    filename = "input_5.txt"
    passes::Vector{String} = readInput(filename)

    return passes
end

function decode(bp::String, bins::Int=8)
    row = [0, 2^bins-1]
    for i in 1:bins-1
        if bp[i] == 'F' || bp[i] == 'L'
            row[2] = 2^*(bins-i) * 0.5-1 + row[1]
        elseif bp[i] == 'B' || bp[i] == 'R'
            row[1] = 2^*(bins-i) * 0.5 + row[1]
        end
        #println(row)
    end
    return row[1]
end

function part1(passes::Vector{String})
    ids = Vector{Int}()
    for bp in passes
        row = decode(bp[1:7])
        seat = decode(bp[8:end], 4)
        id = row*8 + seat
        push!(ids, id)
    end
    println("🎄 Part 1: $(maximum(ids))")
    return ids
end

function part2(ids::Vector{Int})
    sort!(ids)
    r = minimum(ids):maximum(ids)
    miss = setdiff(Set(r), Set(ids))
    println("🎄🎅 Part 2: $(pop!(miss))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 5         ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()
@time ids = part1(inputs)
@time part2(ids)

File: 2020/5.py
import time
from utils import readInput

def loadInput():
    lines = readInput("input_5.txt")

    return lines

def rule(code, up=128):
    diff = up // 2
    for c in code:
        if c == 'F' or c == 'L':
            up -= diff
        diff //= 2
    return up - 1

def part1(lines):
    #code = "BFFFBBFRRR" #: row 70, column 7, seat ID 567.
    #code = "FFFBBBFRRR" #: row 14, column 7, seat ID 119.
    #code = "BBFFBBFRLL" #:
    #code = "FBFBBFFRLR"

    maxID = 0
    for l in lines:
        row = rule(l[:7])
        col = rule(l[7:], 8)
        id  = row*8+col
        if id > maxID:
            maxID = id

    print ("🎄 Part 1: {}".format(maxID))

def part2(lines):
    ids = []
    seats  = []
    for l in lines:
        row = rule(l[:7])
        col = rule(l[7:], 8)
        ids.append(row*8+col)
        seats.append((col, row))

    ids = set(ids)
    all = set(range(59, 905))

    for y in range(128):
        print ("{:3} ".format(y), end='')
        for x in range(8):
            if (x, y) in seats:
                print ("X", end="")
            else:
                print (".", end="")
            if x == 3:
                print (" ", end="")
        print ()

    print ("🎄🎅 Part 2: {}".format(all-ids))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 5         ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()

t0 = time.time()
part1(inputs)
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2(inputs)
print ("Time: {:.5f}".format(time.time()-t0))
